realclientip_test.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetRealClientIPParser(t *testing.T) {
+	var p realClientIPParser
+	var err error
+
+	p, err = getRealClientIPParser("X-Forwarded-For")
+	assert.Nil(t, err)
+	assert.Equal(t, reflect.TypeOf(p), reflect.TypeOf((*xForwardedForClientIPParser)(nil)))
+
+	p, err = getRealClientIPParser("X-REAL-IP")
+	assert.Nil(t, err)
+	assert.Equal(t, reflect.TypeOf(p), reflect.TypeOf((*xForwardedForClientIPParser)(nil)))
+	if xp, ok := p.(*xForwardedForClientIPParser); ok {
+		assert.Equal(t, xp.header, http.CanonicalHeaderKey("X-Real-Ip"))
+	} else {
+		panic("Type of local variable p changed without assignment?")
+	}
+
+	p, err = getRealClientIPParser("x-proxyuser-ip")
+	assert.Nil(t, err)
+	assert.Equal(t, reflect.TypeOf(p), reflect.TypeOf((*xForwardedForClientIPParser)(nil)))
+
+	p, err = getRealClientIPParser("")
+	assert.NotNil(t, err)
+	assert.Equal(t, err.Error(), "The HTTP header key () is either invalid or unsupported")
+	assert.Nil(t, p)
+
+	p, err = getRealClientIPParser("Forwarded")
+	assert.NotNil(t, err)
+	assert.Equal(t, err.Error(), "The HTTP header key (Forwarded) is either invalid or unsupported")
+	assert.Nil(t, p)
+
+	p, err = getRealClientIPParser("2#* @##$$:kd")
+	assert.NotNil(t, err)
+	assert.Equal(t, err.Error(), "The HTTP header key (2#* @##$$:kd) is either invalid or unsupported")
+	assert.Nil(t, p)
+}
+
+func TestXForwardedForClientIPParser(t *testing.T) {
```

Comment:
This might also be better as a table with some descriptions of what each of the cases we are checking are, I think this would certainly help me to understand what's going on with each test case and the expected outcome

Reply:
Agreed, will do.

Reply:
Is it easier to follow the test cases now?